pintracker/statelesstracker/statelesstracker.go

```diff
@@ -0,0 +1,510 @@
+			}
+		case <-spt.ctx.Done():
+			return
+		}
+	}
+}
+
+// reads the queue and makes unpin requests to the IPFS daemon
+func (spt *Tracker) unpinWorker() {
+	for {
+		select {
+		case p := <-spt.unpinCh:
+			if opc, ok := spt.optracker.Get(p.Cid); ok && opc.Op == optracker.OperationUnpin {
+				spt.optracker.UpdateOperationPhase(p.Cid, optracker.PhaseInProgress)
+				spt.unpin(p)
+			}
+		case <-spt.ctx.Done():
+			return
+		}
+	}
+}
+
+// SetClient makes the StatelessPinTracker ready to perform RPC requests to
+// other components.
+func (spt *Tracker) SetClient(c *rpc.Client) {
+	spt.rpcClient = c
+	spt.rpcReady <- struct{}{}
+}
+
+// Shutdown finishes the services provided by the StatelessPinTracker
+// and cancels any active context.
+func (spt *Tracker) Shutdown() error {
+	spt.shutdownMu.Lock()
+	defer spt.shutdownMu.Unlock()
+
+	if spt.shutdown {
+		logger.Debug("already shutdown")
+		return nil
+	}
+
+	logger.Info("stopping StatelessPinTracker")
+	spt.cancel()
+	close(spt.rpcReady)
+	spt.wg.Wait()
+	spt.shutdown = true
+	return nil
+}
+
+// Track tells the StatelessPinTracker to start managing a Cid,
+// possibly triggering Pin operations on the IPFS daemon.
+func (spt *Tracker) Track(c api.Pin) error {
+	logger.Debugf("tracking %s", c.Cid)
+	if opc, ok := spt.optracker.Get(c.Cid); ok {
+		if opc.Op == optracker.OperationUnpin {
+			switch opc.Phase {
+			case optracker.PhaseQueued:
+				spt.optracker.Finish(c.Cid)
+				return nil
+			case optracker.PhaseInProgress:
+				spt.optracker.Finish(c.Cid)
+				// NOTE: this may leave the api.PinInfo in an error state
+				// so a pin operation needs to be run on it (same as Recover)
+			}
+		}
+	}
+
+	if util.IsRemotePin(c, spt.peerID) {
+		if spt.Status(c.Cid).Status == api.TrackerStatusPinned {
+			spt.optracker.TrackNewOperation(
+				spt.ctx,
+				c.Cid,
+				optracker.OperationUnpin,
+			)
+			spt.unpin(c)
+		}
+		spt.setRemote(c.Cid)
+		return nil
+	}
+	spt.removeRemote(c.Cid)
+
+	spt.optracker.TrackNewOperation(spt.ctx, c.Cid, optracker.OperationPin)
+
+	select {
+	case spt.pinCh <- c:
+	default:
+		err := errors.New("pin queue is full")
+		spt.setError(c.Cid, err)
+		logger.Error(err.Error())
+		return err
+	}
+	return nil
+}
+
+// Untrack tells the StatelessPinTracker to stop managing a Cid.
+// If the Cid is pinned locally, it will be unpinned.
+func (spt *Tracker) Untrack(c *cid.Cid) error {
+	logger.Debugf("untracking %s", c)
+	if opc, ok := spt.optracker.Get(c); ok {
+		if opc.Op == optracker.OperationPin {
+			spt.optracker.Finish(c) // cancel it
+
+			switch opc.Phase {
+			case optracker.PhaseQueued:
+				return nil
+			case optracker.PhaseInProgress:
+				// continues below to run a full unpin
+			}
+		}
+	}
+
+	spt.optracker.TrackNewOperation(spt.ctx, c, optracker.OperationUnpin)
+
+	select {
+	case spt.unpinCh <- api.PinCid(c):
+	default:
+		err := errors.New("unpin queue is full")
+		spt.setError(c, err)
+		logger.Error(err.Error())
+		return err
+	}
+	return nil
+}
```

Comment:
Note: `Track` and `Untrack` are almost identical. First, if it's code doing exactly the same they should be identical. Second, probably worth refactoring. Perhaps it's worth having a `base` tracker that implements partial functionality for both since they share a lot of code.